torch/distributed/tensor/_sharding_prop.py

```diff
@@ -48,6 +54,102 @@ def cache_info(self):
         return self.cache.cache_info()
 
 
+# This is a fast-path for sharding prop. The idea is as follows:
+# (1) normally, sharding prop is "smarter". It proposes several different sharding strategies,
+#     and tries all of them to find the optimal strategy. Some consequences:
+#      (a) these strategies can add collectives to get to the final output sharding
+#      (b) these strategies can be dependent on the input shape
+# (2) In basic TP/EP situations, we can do better:
+#     (a) the sharding strategy is "obvious": e.g. mm(replicate, shard(1)) -> shard(1)
+#     (b) the known output sharding does not require emiting any extra collectives inside DTensor
+#     (c) the decided output sharding is agnostic to the input shapes
+# This has some benefits for both compile and eager:
+# - in eager, we can skip expensive sharding prop, which would normally not be cacheable in cases
+#   where input shapes are constantly changing due to data dependencies
+# - in compile, if our shapes are data dependent and therefore not known at compile time,
+#   we wouldn't even be *able* to use the more complicated, shape-dependent strategies
+def _propagate_op_sharding_fast_path(op_info):
+    from torch._subclasses import FakeTensorMode
+
+    out_placements: Optional[tuple[Placement]] = None
+    if op_info.schema.op is torch.ops.aten.mm.default:
```

Comment:
I am a bit worrying that the special rule we need to do to support dynamic shape would be too much to handle.. If we want to do this, at least we should create another dict that map <op, fast_path_rule> instead of manually if checks.